src/repo.rs

```diff
@@ -1318,14 +1360,15 @@ impl Repo {
                 sink,
             } => {
                 if let Some(RemoteRepo {
-                    stream: existing_stream,
+                    stream: _existing_stream,
                     sink: existing_sink,
                 }) = self.remote_repos.remove(&repo_id)
                 {
+                    tracing::debug!(remote_repo=?repo_id, "replacing existing remote repo with new connection");
                     // Reset the sync state.
                     self.remove_unused_sync_states();
                     let remote = RemoteRepo {
-                        stream: Box::new(existing_stream.chain(stream)),
```

Comment:
I think we should explicitly _not_ chain streams for the following reason. The contract of the sync protocol is that the user must provide a reliable in-order stream and _reset state on reconnect_. This last part is very important because without it you can end up looping or stalling. 

The situation this code is handling, where there are several streams connected to the same peer is incompatible with the reliable in-order stream guarantee, which is why we drop the old stream and use the new one and it's also why I think we should not pull any more messages out of the existing stream or send any more messages to the existing sink. We need the reset of the sync state and the switch to new stream/sink to all happen at once to maintain the reliable in-order guarantee. 

I actually had to do this because some of the tests were failing after switching to the `tincans` because now that the sink and stream were separate things (with separate lifetimes) the repo was pulling old messages out of the streams even after the sink was closed and confusing the sync protocol.

Reply:
Ok, thanks for the explanation.